File: 2022/hs/day-25.hs
import Data.Foldable (foldl')
import qualified Data.Bimap as Bimap
import Data.Bimap ((!), (!>))

main = do
    input <- readFile "inputs/25.txt"
    let digits = [fmap (snafuDigit !) line | line <- lines input]

    putStr "sum: "
    print $ sum $ fmap (fromBase 5) digits
    
    putStr "sum-snafu: "
    putStrLn $ toSnafu $ sum $ fmap (fromBase 5) digits

snafuDigit = Bimap.fromList [
    ('2', 2),
    ('1', 1),
    ('0', 0),
    ('-', -1),
    ('=', -2)
    ]

fromBase n = foldl' (\a b -> a * n + b) 0

balQR n d = if r > (d `div` 2) then (q + 1, r - d) else (q, r)
    where (q, r) = quotRem n d

toSnafu 0 = "0"
toSnafu n = conv n []
    where 
        conv 0 = id
        conv n = conv q . ([snafuDigit !> r] ++)
            where (q, r) = balQR n 5

File: 2022/src/bin/day-25.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("inputs/25.txt").unwrap();

    let mut sum = 0;
    for line in input.lines() {
        let val = from_snafu(line);
        println!("{line} => {val}");
        sum += val;
    }
    println!("sum: {sum}");
    println!("sum-snafu: {}", to_snafu(sum));
}

fn from_snafu(s: &str) -> isize {
    let mut result = 0;
    for c in s.chars() {
        result *= 5;
        result += match c {
            '2' => 2,
            '1' => 1,
            '0' => 0,
            '-' => -1,
            '=' => -2,
            _ => unimplemented!()
        }
    }

    result
}

fn bal_mod_5(val: isize) -> isize {
    let v5 = val.rem_euclid(5);
    if v5 > 2 { v5 - 5 } else { v5 }
}

fn to_snafu(mut val: isize) -> String {
    let mut vec = vec![];

    while val != 0 {
        let m = bal_mod_5(val);
        let c = match m {
            2 => '2',
            1 => '1',
            0 => '0',
            -1 => '-',
            -2 => '=',
            _ => unimplemented!()
        };
        vec.push(c);

        val -= m;
        val /= 5;
    }

    vec.into_iter().rev().collect()
}